_stbt/motion.py

```diff
@@ -0,0 +1,348 @@
+
+        absdiff = cv2.absdiff(frame_gray, previous_frame_gray)
+        previous_frame_gray = frame_gray
+        imglog.imwrite("absdiff", absdiff)
+
+        if mask.image is not None:
+            absdiff = cv2.bitwise_and(absdiff, mask.image)
+            imglog.imwrite("mask", mask.image)
+            imglog.imwrite("absdiff_masked", absdiff)
+
+        _, thresholded = cv2.threshold(
+            absdiff, int((1 - noise_threshold) * 255), 255,
+            cv2.THRESH_BINARY)
+        eroded = cv2.erode(
+            thresholded,
+            cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (3, 3)))
+        imglog.imwrite("absdiff_threshold", thresholded)
+        imglog.imwrite("absdiff_threshold_erode", eroded)
+
+        out_region = _pixel_bounding_box(eroded)
+        if out_region:
+            # Undo cv2.erode above:
+            out_region = out_region.extend(x=-1, y=-1)
+            # Undo crop:
+            out_region = out_region.translate(region.x, region.y)
+
+        motion = bool(out_region)
+
+        result = MotionResult(getattr(frame, "time", None), motion,
+                              out_region, frame)
+        draw_on(frame, result, label="detect_motion()")
+        debug("%s found: %s" % (
+            "Motion" if motion else "No motion", str(result)))
+        yield result
+
+
+def _pixel_bounding_box(img):
+    """
+    Find the smallest region that contains all the non-zero pixels in an image.
+
+    >>> _pixel_bounding_box(numpy.array([[0]], dtype=numpy.uint8))
+    >>> _pixel_bounding_box(numpy.array([[1]], dtype=numpy.uint8))
+    Region(x=0, y=0, right=1, bottom=1)
+    >>> _pixel_bounding_box(numpy.array([
+    ...     [0, 0, 0, 0],
+    ...     [0, 1, 1, 1],
+    ...     [0, 1, 1, 1],
+    ...     [0, 0, 0, 0],
+    ... ], dtype=numpy.uint8))
+    Region(x=1, y=1, right=4, bottom=3)
+    >>> _pixel_bounding_box(numpy.array([
+    ...     [0, 0, 0, 0, 0, 0],
+    ...     [0, 0, 0, 1, 0, 0],
+    ...     [0, 1, 0, 0, 0, 0],
+    ...     [0, 0, 0, 0, 1, 0],
+    ...     [0, 0, 1, 0, 0, 0],
+    ...     [0, 0, 0, 0, 0, 0]
+    ... ], dtype=numpy.uint8))
+    Region(x=1, y=1, right=5, bottom=5)
+    """
+    if len(img.shape) != 2:
+        raise ValueError("Single-channel image required.  Provided image has "
+                         "shape %r" % (img.shape,))
+
+    out = [None, None, None, None]
+
+    for axis in (0, 1):
+        flat = numpy.any(img, axis=axis)
+        indices = numpy.where(flat)[0]
+        if len(indices) == 0:
+            return None
+        out[axis] = indices[0]
+        out[axis + 2] = indices[-1] + 1
+
+    return Region.from_extents(*out)
+
+
+def limit_time(frames, duration_secs):
+    """
+    Adapts a frame iterator such that it will return EOS after `duration_secs`
+    worth of video has been read.
+    """
+    end_time = None
+    for frame, ts in frames:
+        if end_time is None:
+            end_time = frame.time + duration_secs
```

Comment:
This is a change in behaviour: The timeout is the first frame's time + 10s, instead of `time.time()` + 10s.

This means that the effective timeout will be slightly shorter than before.

Reply:
Fixed.